File: tests/algorithms/x/Clojure/maths/numerical_analysis/bisection.clj
(ns main (:refer-clojure :exclude [abs_float bisection f main]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(defn toi [s]
  (Integer/parseInt (str s)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare abs_float bisection f main)

(def ^:dynamic bisection_end nil)

(def ^:dynamic bisection_fmid nil)

(def ^:dynamic bisection_mid nil)

(def ^:dynamic bisection_start nil)

(defn abs_float [abs_float_x]
  (try (if (< abs_float_x 0.0) (throw (ex-info "return" {:v (- abs_float_x)})) (throw (ex-info "return" {:v abs_float_x}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn bisection [f_v bisection_a bisection_b]
  (binding [bisection_end nil bisection_fmid nil bisection_mid nil bisection_start nil] (try (do (set! bisection_start bisection_a) (set! bisection_end bisection_b) (when (= (f_v bisection_a) 0.0) (throw (ex-info "return" {:v bisection_a}))) (when (= (f_v bisection_b) 0.0) (throw (ex-info "return" {:v bisection_b}))) (when (> (* (f_v bisection_a) (f_v bisection_b)) 0.0) (throw (Exception. "could not find root in given interval."))) (set! bisection_mid (+ bisection_start (/ (- bisection_end bisection_start) 2.0))) (while (> (abs_float (- bisection_start bisection_mid)) 0.0000001) (do (set! bisection_fmid (f_v bisection_mid)) (when (= bisection_fmid 0.0) (throw (ex-info "return" {:v bisection_mid}))) (if (< (* bisection_fmid (f_v bisection_start)) 0.0) (set! bisection_end bisection_mid) (set! bisection_start bisection_mid)) (set! bisection_mid (+ bisection_start (/ (- bisection_end bisection_start) 2.0))))) (throw (ex-info "return" {:v bisection_mid}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn f [f_x]
  (try (throw (ex-info "return" {:v (- (- (* (* f_x f_x) f_x) (* 2.0 f_x)) 5.0)})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn main []
  (println (str (bisection f 1.0 1000.0))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (main)
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)

File: tests/algorithms/x/PHP/maths/numerical_analysis/bisection.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function abs_float($x) {
  if ($x < 0.0) {
  return -$x;
} else {
  return $x;
}
};
  function bisection($f, $a, $b) {
  $start = $a;
  $end = $b;
  if ($f($a) == 0.0) {
  return $a;
}
  if ($f($b) == 0.0) {
  return $b;
}
  if ($f($a) * $f($b) > 0.0) {
  $panic('could not find root in given interval.');
}
  $mid = $start + ($end - $start) / 2.0;
  while (abs_float($start - $mid) > 0.0000001) {
  $fmid = $f($mid);
  if ($fmid == 0.0) {
  return $mid;
}
  if ($fmid * $f($start) < 0.0) {
  $end = $mid;
} else {
  $start = $mid;
}
  $mid = $start + ($end - $start) / 2.0;
};
  return $mid;
};
  function f($x) {
  return $x * $x * $x - 2.0 * $x - 5.0;
};
  function main() {
  echo rtrim(_str(bisection('f', 1.0, 1000.0))), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
